test/common/http/http2/codec_impl_test.cc

```diff
@@ -232,42 +235,99 @@ TEST_P(Http2CodecImplTest, TrailingHeadersLargeBody) {
   response_encoder_->encodeTrailers(TestHeaderMapImpl{{"trailing", "header"}});
 }
 
-TEST_P(Http2CodecImplTest, DeferredReset) {
+class Http2CodecImplDeferredResetTest : public Http2CodecImplTest {};
+
+TEST_P(Http2CodecImplDeferredResetTest, DeferredResetClient) {
   InSequence s;
 
-  // Buffer server data so we can make sure we don't get any window updates.
+  // Do a basic request/response to flush settings.
+  TestHeaderMapImpl request_headers;
+  HttpTestUtility::addDefaultHeaders(request_headers);
+  EXPECT_CALL(request_decoder_, decodeHeaders_(_, true));
+  request_encoder_.encodeHeaders(request_headers, true);
+  TestHeaderMapImpl response_headers{{":status", "200"}};
+  EXPECT_CALL(response_decoder_, decodeHeaders_(_, true));
+  response_encoder_->encodeHeaders(response_headers, true);
+
+  // Do a 2nd request, but pause server dispatch so we don't send window updates. This will
+  // result in a deferred reset, followed by a flush which will cause the stream to actually
+  // be reset.
   ON_CALL(client_connection_, write(_))
       .WillByDefault(
           Invoke([&](Buffer::Instance& data) -> void { server_wrapper_.buffer_.add(data); }));
+  StreamEncoder& request_encoder2 = client_.newStream(response_decoder_);
+  request_encoder2.encodeHeaders(request_headers, false);
+  Buffer::OwnedImpl body(std::string(1024 * 1024, 'a'));
+  request_encoder2.encodeData(body, true);
+  request_encoder2.getStream().resetStream(StreamResetReason::LocalReset);
+
+  // Dispatch server. We expect to see some data.
+  expectServerStream();
+  EXPECT_CALL(request_decoder_, decodeHeaders_(_, false))
+      .WillOnce(InvokeWithoutArgs([&]() -> void {
+        // Start a response inside the headers callback. This should not result in the client
+        // seeing any headers as the stream should already be reset on the other side.
```

Comment:
Implicitly it is because of the `EXPECT_CALL(response_decoder_, decodeHeaders_(_, true));` expectation above. I will add a comment.